echo_server.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+
+"""
+Echo Server
+"""
+
+import socket
+import email.utils
+
+
+BUFFERSIZE = 32
+
+
+def echo_server():
+    server_socket = socket.socket(
+        socket.AF_INET,
+        socket.SOCK_STREAM,
+        socket.IPPROTO_IP)
+    server_socket.bind(('127.0.0.1', 50000))
+    server_socket.listen(1)
+
+    try:
+        while True:
+            request = ''
+            done = False
+            conn, addr = server_socket.accept()
+            while not done:
+                msg_part = conn.recv(BUFFERSIZE)
+                if len(msg_part) < BUFFERSIZE:
+                    done = True
+                request += msg_part
+            try:
+                parsed_request = parse_request(request)
+                response = response_ok()
+            except HTTPError400:
+                response = response_error(400, 'HTTP Request malformed.')
+            except HTTPError405:
+                response = response_error(405, 'Not a GET request.')
+            except HTTPError505:
+                response = response_error(505, 'Not HTTP/1.1 protocal.')
+            conn.sendall(response)
+            conn.close()
+    except KeyboardInterrupt:
+        pass
+    server_socket.close()
+
+
+def response_ok(body=''):
+    """return a well formed HTTP "200 OK" response as a byte string
+    """
+    lines = [
+        "HTTP/1.1 200 OK",
+        "Date : {}".format(email.utils.formatdate(usegmt=True)),
+        "Content-Type: text/xml; charset=utf-8",
+        "<html><body><h1>Successful response.</h1></body></html>"
+    ]
+
+    return "".join(["\r\n".join(lines), "\r\n\r\n"])
+
+
+def response_error(code=400, message="Bad Request"):
+    """return a well formed HTTP error response"""
+    error = " ".join([str(code), message])
+    lines = [
+        "HTTP/1.1 {}".format(error),
+        "Date : {}".format(email.utils.formatdate(usegmt=True)),
+        "Content-Type: text/xml; charset=utf-8",
+        "<html><body><h1> {} </h1></body></html>".format(error)
+    ]
+
+    return "".join(["\r\n".join(lines), "\r\n\r\n"])
```

Comment:
same comment as for the previous method.  Missing a blank line between the headers and the response body.